### PR TITLE
Added documentation to Open Alpha Tensor modules and methods

### PR DESCRIPTION
**Pull request by BrianPulfer**
Solving issue #138 .

### CODE REVIEW

**Review by diegofiori — CHANGES_REQUESTED** [review]
Hello @BrianPulfer! Thanks for your contribution! You made a really good job! I just left few comments about a couple of argument descriptions. The majority of my comments are just about adapting the documentation to the style we used on the main repository, `nebullvm`. 

Furthermore, please, check out the [Coding Style section in the Contribution Guidelines](https://github.com/nebuly-ai/nebullvm/blob/main/CONTRIBUTING.md#coding-style) and make sure you have configured your environment correctly (I saw a couple of changes in the code format).

**Review by diegofiori — COMMENTED** [review]
Hello @BrianPulfer, I left a couple of minor comments. Just one thing, be sure to have properly installed the pre-commit library in your environment and to have it properly configured following [these guidelines](https://github.com/nebuly-ai/nebullvm/blob/main/CONTRIBUTING.md#coding-style).

**Comment by diegofiori** [comment]
@BrianPulfer I saw that you resolved my previous comments, but I don't see any modification into the code. Did you push it?

**Comment by BrianPulfer** [comment]
> @BrianPulfer I saw that you resolved my previous comments, but I don't see any modification into the code. Did you push it?

Hi @diegofiori! Yes, sorry about that, I still did not commit those changes yet because of an issue in my environment regarding the pre-commit hooks. Could it be that there's an error in the [.pre-commit-config.yaml](https://github.com/nebuly-ai/nebullvm/blob/main/.pre-commit-config.yaml ) file? It seems to point to flake8's GitLab repo (which doesn't exist). Is this a mistake?

**Comment by diegofiori** [comment]
Hello @BrianPulfer, I pushed a couple of fixes for preventing an excessive memory consumption during training. It creates a couple of conflicts with your code, but they should be quick to be solved. I updated the pre-commit config file too. So, now you should be able to install the hook correctly.

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 1]
```diff
@@ -34,24 +44,25 @@ def remove_duplicates(reducing_tensor: torch.Tensor):
...
         for second_idx in values:
-            old_idx_to_new_idx_map[second_idx] = new_idx
-    return (
-        reducing_tensor[:, indexes],
-        old_idx_to_new_idx_map,
-        idx_map,
-        indexes,
-    )
+            cloned_idx_to_idx_map[second_idx] = key
+    return reducing_tensor[:, indexes], cloned_idx_to_idx_map, indexes
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:51]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 2]
```diff
@@ -65,17 +76,14 @@ def extract_children_states_from_actions(
...
     reducing_tensor = u * v * w
-    (
-        reducing_tensor,
-        old_idx_to_new_idx,
-        repetition_map,
-        not_duplicate_indexes,
-    ) = remove_duplicates(reducing_tensor)
+    reducing_tensor, repetition_map, duplicate_indexes = remove_duplicates(
+        reducing_tensor
+    )
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:86]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 3]
```diff
@@ -91,9 +99,8 @@ def extract_children_states_from_actions(
             )
             for i in range(k)
         ],
-        old_idx_to_new_idx,
         repetition_map,
-        not_duplicate_indexes,
+        duplicate_indexes,
     )
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:104]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 4]
```diff
@@ -125,19 +150,15 @@ def select_future_state(
     possible_states: List[torch.Tensor],
     q_values: torch.Tensor,
     N_s_a: torch.Tensor,
-    repetitions: Dict[int, list],
+    repetitions: Dict[int, int],
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:153]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 5]
```diff
@@ -125,19 +150,15 @@ def select_future_state(
...
 ) -> torch.Tensor:
     """Select the future state maximizing the upper confidence bound."""
     # q_values (1, K, 1)
     pi = torch.tensor(
-        [
-            len(repetitions[i])
-            for i in range(len(possible_states))
-            if i in repetitions
-        ]
+        [repetitions.get(i, 0) for i in range(len(possible_states))]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:161]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 6]
```diff
@@ -161,23 +182,30 @@ def simulate_game(
...
     idx = t_time
     max_steps = min(max_steps, t_time + horizon)
     state_hash = to_hash(extract_present_state(state))
     trajectory = []
     # selection
     while state_hash in game_tree:
         (
-            (state_tensor, full_actions),
-            old_idx_to_new_idx,
+            possible_states,
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:203]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 7]
```diff
@@ -161,23 +182,30 @@ def simulate_game(
...
             repetition_map,
             N_s_a,
             q_values,
             actions,
         ) = states_dict[state_hash]
-        possible_states = extract_children_states_from_actions(
-            state_tensor, full_actions
-        )[0]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:180]: please keep the old code modification (the code in red should be inserted back)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 8]
```diff
@@ -196,7 +224,6 @@ def simulate_game(
             actions, probs, q_values = model(state, scalars)
             (
                 possible_states,
-                cloned_idx_to_idx,
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:199]: please keep the old code modification (the code in red should be inserted back)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 9]
```diff
@@ -210,8 +237,7 @@ def simulate_game(
             N_s_a = torch.zeros_like(not_dupl_q_values).to("cpu")
             present_state = extract_present_state(state)
             states_dict[to_hash(present_state)] = (
-                (state, actions),
-                cloned_idx_to_idx,
+                [s.to("cpu") for s in possible_states],
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:240]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 10]
```diff
@@ -235,19 +261,14 @@ def backward_pass(trajectory, states_dict, leaf_q_value: torch.Tensor):
...
-                _,
-                old_idx_to_new_idx,
-                _,
-                N_s_a,
-                q_values,
-                _,
-            ) = states_dict[state]
+            possible_states, repetitions, N_s_a, q_values, _ = states_dict[
+                state
+            ]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:266]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 11]
```diff
@@ -235,19 +261,14 @@ def backward_pass(trajectory, states_dict, leaf_q_value: torch.Tensor):
...
             if isinstance(reward, torch.Tensor):
                 reward = reward.to(q_values.device)
             action_idx = int(action_idx)
-            if action_idx in old_idx_to_new_idx:
-                not_dupl_index = old_idx_to_new_idx[int(action_idx)]
+            if action_idx in repetitions:
+                not_dupl_index = repetitions[int(action_idx)]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:271]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 12]
```diff
@@ -274,19 +306,14 @@ def monte_carlo_tree_search(
     state_hash = to_hash(extract_present_state(state))
     if state_hash in state_dict:
         with torch.no_grad():
-            N_s_a = state_dict[state_hash][3]
+            N_s_a = state_dict[state_hash][2]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:309]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 13]
```diff
@@ -274,19 +306,14 @@ def monte_carlo_tree_search(
...
     for _ in range(n_sim):
         simulate_game(model, state, t_time, n_steps, game_tree, state_dict)
     # return next state
-    (state_tensor, actions), _, repetitions, N_s_a, q_values, _ = state_dict[
-        state_hash
-    ]
-    possible_states = extract_children_states_from_actions(
-        state_tensor, actions
-    )[0]
+    possible_states, repetitions, N_s_a, q_values, _ = state_dict[state_hash]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:316]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 14]
```diff
@@ -302,15 +329,15 @@ def compute_improved_policy(
...
+    The improved policy is computed as (N_s_a / N_s_a.sum())ˆ(1/tau) where tau
+    is (log(N_s_a.sum()) / log(N_bar))
     """
     policies = torch.zeros(len(states), model_n_steps, model_n_logits)
     N_bar = torch.tensor(N_bar)
     for idx, state in enumerate(states):
-        N_s_a = state_dict[state][3]
-        actions = state_dict[state][5]
+        N_s_a = state_dict[state][2]
+        actions = state_dict[state][4]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:340]: please keep the old code modification (the code in red should replace the code in green)

File: apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py [hunk 15]
```diff
@@ -368,7 +406,7 @@ def actor_prediction(
         torch.tensor([-1] * (len(policies) - 1) + [reward]), dim=0
     )
     if return_actions:
-        actions = [state_dict[hash_state][5] for hash_state in hash_states]
+        actions = [state_dict[hash_state][4] for hash_state in hash_states]
```
Comment (diegofiori) [inline apps/accelerate/open_alpha_tensor/open_alpha_tensor/core/actors/stage.py:409]: please keep the old code modification (the code in red should replace the code in green)

**Review by diegofiori — COMMENTED** [review]
Thanks for the contribution @BrianPulfer! I think there have been a couple of problems while fixing the merge conflicts. I just left few comments where the modifications need to be reverted.

**Comment by BrianPulfer** [comment]
Hi @diegofiori. I should have figured it out. Sorry for the unnecessarily many commits 😅. It should all be good now.

**Review by diegofiori — APPROVED** [review]
Thanks @BrianPulfer for your contribution! Very much appreciated. Since your contribution was super valuable, I'm going to send over some swags from the community. Also, I would be very happy to work together on adding more features to the `open_alpha_tensor` module. 

I was planning to add the following features:
- [ ] Action to Code conversion: Convert the model outputs (sequence of actions) into a code for solving matrix multiplication
- [ ] Function for running the AlphaTensor model at inference time, using n actors working on n different basis
- [ ] Convert results obtained in different basis to the original one
- [ ] Improve the speed of the acting step, which is right now the main bottleneck while training the model.

Let me know if you are interested in implementing any of them!

**Pull request merged**